File: WebUI/src/utils/binaryProtocol.ts
// Ultra-fast binary protocol client for maximum performance
// Eliminates JSON parsing overhead for real-time memory monitoring

export enum BinaryOpcode {
  MEMORY_READ = 0x01,
  MEMORY_WRITE = 0x02,
  MEMORY_SCAN = 0x03,
  VALUE_UPDATE = 0x04,
  BULK_UPDATE = 0x05,
  PROCESS_INFO = 0x06,
  MODULE_LIST = 0x07,
  PING = 0x08,
  PONG = 0x09
}

export enum DataType {
  INT32 = 0x01,
  INT64 = 0x02,
  FLOAT = 0x03,
  DOUBLE = 0x04,
  STRING = 0x05,
  BYTES = 0x06
}

export interface BinaryHeader {
  magic: number // 0x494E544C
  version: number // 0x0001
  opcode: BinaryOpcode
  flags: number
  payloadSize: number
  requestId: number
}

export interface MemoryReadRequest {
  address: bigint
  size: number
  type: DataType
}

export interface MemoryWriteRequest {
  address: bigint
  type: DataType
  dataSize: number
  data: Uint8Array
}

export interface ValueUpdate {
  address: bigint
  type: DataType
  value: Uint8Array
}

export class BinaryProtocolClient {
  private requestIdCounter = 1
  private pendingRequests = new Map<number, (response: any) => void>()
  
  // Encode binary messages
  encodeMemoryRead(address: string, size: number, type: string): ArrayBuffer {
    const requestId = this.requestIdCounter++
    const buffer = new ArrayBuffer(32) // Header + MemoryReadRequest
    const view = new DataView(buffer)
    
    // Header
    view.setUint32(0, 0x494E544C, true) // magic
    view.setUint16(4, 0x0001, true) // version
    view.setUint8(6, BinaryOpcode.MEMORY_READ)
    view.setUint8(7, 0) // flags
    view.setUint32(8, 16, true) // payload size
    view.setUint32(12, requestId, true)
    
    // Payload
    const addr = BigInt(address)
    view.setBigUint64(16, addr, true)
    view.setUint32(24, size, true)
    view.setUint8(28, this.stringToDataType(type))
    
    return buffer
  }
  
  encodeMemoryWrite(address: string, type: string, data: Uint8Array): ArrayBuffer {
    const requestId = this.requestIdCounter++
    const buffer = new ArrayBuffer(32 + data.length) // Header + MemoryWriteRequest + data
    const view = new DataView(buffer)
    
    // Header
    view.setUint32(0, 0x494E544C, true) // magic
    view.setUint16(4, 0x0001, true) // version
    view.setUint8(6, BinaryOpcode.MEMORY_WRITE)
    view.setUint8(7, 0) // flags
    view.setUint32(8, 16 + data.length, true) // payload size
    view.setUint32(12, requestId, true)
    
    // Payload
    const addr = BigInt(address)
    view.setBigUint64(16, addr, true)
    view.setUint8(24, this.stringToDataType(type))
    view.setUint32(25, data.length, true)
    
    // Data
    const uint8View = new Uint8Array(buffer, 29)
    uint8View.set(data)
    
    return buffer
  }
  
  // Decode binary messages
  decodeHeader(buffer: ArrayBuffer): BinaryHeader | null {
    if (buffer.byteLength < 16) return null
    
    const view = new DataView(buffer)
    const magic = view.getUint32(0, true)
    
    if (magic !== 0x494E544C) return null
    
    return {
      magic,
      version: view.getUint16(4, true),
      opcode: view.getUint8(6),
      flags: view.getUint8(7),
      payloadSize: view.getUint32(8, true),
      requestId: view.getUint32(12, true)
    }
  }
  
  decodeValueUpdate(buffer: ArrayBuffer, offset: number = 16): ValueUpdate | null {
    if (buffer.byteLength < offset + 16) return null
    
    const view = new DataView(buffer, offset)
    const address = view.getBigUint64(0, true)
    const type = view.getUint8(8) as DataType
    const valueSize = view.getUint32(9, true)
    
    if (buffer.byteLength < offset + 13 + valueSize) return null
    
    const value = new Uint8Array(buffer, offset + 13, valueSize)
    
    return { address, type, value }
  }
  
  decodeBulkUpdate(buffer: ArrayBuffer, offset: number = 16): ValueUpdate[] {
    const updates: ValueUpdate[] = []
    const view = new DataView(buffer, offset)
    const count = view.getUint32(0, true)
    
    let currentOffset = offset + 4
    for (let i = 0; i < count; i++) {
      const update = this.decodeValueUpdate(buffer, currentOffset)
      if (!update) break
      
      updates.push(update)
      currentOffset += 13 + update.value.length
    }
    
    return updates
  }
  
  // Utility functions
  private stringToDataType(type: string): DataType {
    switch (type) {
      case 'int32': return DataType.INT32
      case 'int64': return DataType.INT64
      case 'float': return DataType.FLOAT
      case 'double': return DataType.DOUBLE
      case 'string': return DataType.STRING
      case 'bytes': return DataType.BYTES
      default: return DataType.BYTES
    }
  }
  
  private dataTypeToString(type: DataType): string {
    switch (type) {
      case DataType.INT32: return 'int32'
      case DataType.INT64: return 'int64'
      case DataType.FLOAT: return 'float'
      case DataType.DOUBLE: return 'double'
      case DataType.STRING: return 'string'
      case DataType.BYTES: return 'bytes'
      default: return 'bytes'
    }
  }
  
  // Convert value bytes to string based on type
  valueToString(value: Uint8Array, type: DataType): string {
    const view = new DataView(value.buffer, value.byteOffset, value.byteLength)
    
    switch (type) {
      case DataType.INT32:
        return view.getInt32(0, true).toString()
      case DataType.INT64:
        return view.getBigInt64(0, true).toString()
      case DataType.FLOAT:
        return view.getFloat32(0, true).toString()
      case DataType.DOUBLE:
        return view.getFloat64(0, true).toString()
      case DataType.STRING:
        return new TextDecoder().decode(value)
      case DataType.BYTES:
        return Array.from(value).map(b => b.toString(16).padStart(2, '0')).join(' ')
      default:
        return ''
    }
  }
  
  // Convert string to value bytes based on type
  stringToValue(str: string, type: DataType): Uint8Array {
    switch (type) {
      case DataType.INT32: {
        const buffer = new ArrayBuffer(4)
        new DataView(buffer).setInt32(0, parseInt(str), true)
        return new Uint8Array(buffer)
      }
      case DataType.INT64: {
        const buffer = new ArrayBuffer(8)
        new DataView(buffer).setBigInt64(0, BigInt(str), true)
        return new Uint8Array(buffer)
      }
      case DataType.FLOAT: {
        const buffer = new ArrayBuffer(4)
        new DataView(buffer).setFloat32(0, parseFloat(str), true)
        return new Uint8Array(buffer)
      }
      case DataType.DOUBLE: {
        const buffer = new ArrayBuffer(8)
        new DataView(buffer).setFloat64(0, parseFloat(str), true)
        return new Uint8Array(buffer)
      }
      case DataType.STRING:
        return new TextEncoder().encode(str)
      case DataType.BYTES:
        return new Uint8Array(str.split(' ').map(b => parseInt(b, 16)))
      default:
        return new Uint8Array()
    }
  }
}

export const binaryProtocol = new BinaryProtocolClient()

File: Internal-Engine/BinaryProtocol.cpp
#include "BinaryProtocol.hpp"
#include <cstring>
#include <algorithm>

namespace InternalEngine {

std::vector<uint8_t> BinaryProtocol::EncodeMemoryRead(uint32_t requestId, uint64_t address, uint32_t size, DataType type) {
    std::vector<uint8_t> message;
    
    // Header
    BinaryHeader header;
    header.opcode = BinaryOpcode::MEMORY_READ;
    header.flags = 0;
    header.payloadSize = sizeof(MemoryReadRequest);
    header.requestId = requestId;
    
    message.resize(sizeof(BinaryHeader) + sizeof(MemoryReadRequest));
    memcpy(message.data(), &header, sizeof(BinaryHeader));
    
    // Payload
    MemoryReadRequest request;
    request.address = address;
    request.size = size;
    request.type = type;
    
    memcpy(message.data() + sizeof(BinaryHeader), &request, sizeof(MemoryReadRequest));
    
    return message;
}

std::vector<uint8_t> BinaryProtocol::EncodeMemoryWrite(uint32_t requestId, uint64_t address, DataType type, const std::vector<uint8_t>& data) {
    std::vector<uint8_t> message;
    
    // Header
    BinaryHeader header;
    header.opcode = BinaryOpcode::MEMORY_WRITE;
    header.flags = 0;
    header.payloadSize = sizeof(MemoryWriteRequest) + static_cast<uint32_t>(data.size());
    header.requestId = requestId;
    
    message.resize(sizeof(BinaryHeader) + sizeof(MemoryWriteRequest) + data.size());
    memcpy(message.data(), &header, sizeof(BinaryHeader));
    
    // Payload
    MemoryWriteRequest request;
    request.address = address;
    request.type = type;
    request.dataSize = static_cast<uint32_t>(data.size());
    
    memcpy(message.data() + sizeof(BinaryHeader), &request, sizeof(MemoryWriteRequest));
    memcpy(message.data() + sizeof(BinaryHeader) + sizeof(MemoryWriteRequest), data.data(), data.size());
    
    return message;
}

std::vector<uint8_t> BinaryProtocol::EncodeValueUpdate(uint64_t address, DataType type, const std::vector<uint8_t>& value) {
    std::vector<uint8_t> message;
    
    // Header
    BinaryHeader header;
    header.opcode = BinaryOpcode::VALUE_UPDATE;
    header.flags = 0;
    header.payloadSize = sizeof(ValueUpdateNotification) + static_cast<uint32_t>(value.size());
    header.requestId = 0; // Not a request
    
    message.resize(sizeof(BinaryHeader) + sizeof(ValueUpdateNotification) + value.size());
    memcpy(message.data(), &header, sizeof(BinaryHeader));
    
    // Payload
    ValueUpdateNotification notification;
    notification.address = address;
    notification.type = type;
    notification.valueSize = static_cast<uint32_t>(value.size());
    
    memcpy(message.data() + sizeof(BinaryHeader), &notification, sizeof(ValueUpdateNotification));
    memcpy(message.data() + sizeof(BinaryHeader) + sizeof(ValueUpdateNotification), value.data(), value.size());
    
    return message;
}

std::vector<uint8_t> BinaryProtocol::EncodeBulkUpdate(const std::vector<std::pair<uint64_t, std::vector<uint8_t>>>& updates) {
    std::vector<uint8_t> message;
    
    // Calculate total payload size
    uint32_t totalPayloadSize = sizeof(BulkUpdateNotification);
    for (const auto& update : updates) {
        totalPayloadSize += sizeof(ValueUpdateNotification) + static_cast<uint32_t>(update.second.size());
    }
    
    // Header
    BinaryHeader header;
    header.opcode = BinaryOpcode::BULK_UPDATE;
    header.flags = 0;
    header.payloadSize = totalPayloadSize;
    header.requestId = 0; // Not a request
    
    message.resize(sizeof(BinaryHeader) + totalPayloadSize);
    memcpy(message.data(), &header, sizeof(BinaryHeader));
    
    // Bulk notification header
    BulkUpdateNotification bulkNotification;
    bulkNotification.count = static_cast<uint32_t>(updates.size());
    
    size_t offset = sizeof(BinaryHeader);
    memcpy(message.data() + offset, &bulkNotification, sizeof(BulkUpdateNotification));
    offset += sizeof(BulkUpdateNotification);
    
    // Individual updates
    for (const auto& update : updates) {
        ValueUpdateNotification notification;
        notification.address = update.first;
        notification.type = DataType::BYTES; // Default to bytes for flexibility
        notification.valueSize = static_cast<uint32_t>(update.second.size());
        
        memcpy(message.data() + offset, &notification, sizeof(ValueUpdateNotification));
        offset += sizeof(ValueUpdateNotification);
        
        memcpy(message.data() + offset, update.second.data(), update.second.size());
        offset += update.second.size();
    }
    
    return message;
}

bool BinaryProtocol::DecodeHeader(const std::vector<uint8_t>& data, BinaryHeader& header) {
    if (data.size() < sizeof(BinaryHeader)) {
        return false;
    }
    
    memcpy(&header, data.data(), sizeof(BinaryHeader));
    
    // Validate magic number
    return header.magic == 0x494E544C;
}

bool BinaryProtocol::DecodeMemoryRead(const std::vector<uint8_t>& payload, MemoryReadRequest& request) {
    if (payload.size() < sizeof(MemoryReadRequest)) {
        return false;
    }
    
    memcpy(&request, payload.data(), sizeof(MemoryReadRequest));
    return true;
}

bool BinaryProtocol::DecodeMemoryWrite(const std::vector<uint8_t>& payload, MemoryWriteRequest& request, std::vector<uint8_t>& data) {
    if (payload.size() < sizeof(MemoryWriteRequest)) {
        return false;
    }
    
    memcpy(&request, payload.data(), sizeof(MemoryWriteRequest));
    
    if (payload.size() < sizeof(MemoryWriteRequest) + request.dataSize) {
        return false;
    }
    
    data.resize(request.dataSize);
    memcpy(data.data(), payload.data() + sizeof(MemoryWriteRequest), request.dataSize);
    
    return true;
}

std::string BinaryProtocol::DataTypeToString(DataType type) {
    switch (type) {
        case DataType::INT32: return "int32";
        case DataType::INT64: return "int64";
        case DataType::FLOAT: return "float";
        case DataType::DOUBLE: return "double";
        case DataType::STRING: return "string";
        case DataType::BYTES: return "bytes";
        default: return "unknown";
    }
}

DataType BinaryProtocol::StringToDataType(const std::string& typeStr) {
    if (typeStr == "int32") return DataType::INT32;
    if (typeStr == "int64") return DataType::INT64;
    if (typeStr == "float") return DataType::FLOAT;
    if (typeStr == "double") return DataType::DOUBLE;
    if (typeStr == "string") return DataType::STRING;
    if (typeStr == "bytes") return DataType::BYTES;
    return DataType::BYTES; // Default fallback
}

uint64_t BinaryProtocol::StringToAddress(const std::string& addressStr) {
    return std::stoull(addressStr, nullptr, 16);
}

std::string BinaryProtocol::AddressToString(uint64_t address) {
    char buffer[32];
    sprintf_s(buffer, "0x%016llX", address);
    return std::string(buffer);
}

void BinaryProtocol::WriteUInt32(std::vector<uint8_t>& buffer, uint32_t value) {
    buffer.push_back(static_cast<uint8_t>(value & 0xFF));
    buffer.push_back(static_cast<uint8_t>((value >> 8) & 0xFF));
    buffer.push_back(static_cast<uint8_t>((value >> 16) & 0xFF));
    buffer.push_back(static_cast<uint8_t>((value >> 24) & 0xFF));
}

void BinaryProtocol::WriteUInt64(std::vector<uint8_t>& buffer, uint64_t value) {
    for (int i = 0; i < 8; i++) {
        buffer.push_back(static_cast<uint8_t>((value >> (i * 8)) & 0xFF));
    }
}

uint32_t BinaryProtocol::ReadUInt32(const uint8_t* data) {
    return static_cast<uint32_t>(data[0]) |
           (static_cast<uint32_t>(data[1]) << 8) |
           (static_cast<uint32_t>(data[2]) << 16) |
           (static_cast<uint32_t>(data[3]) << 24);
}

uint64_t BinaryProtocol::ReadUInt64(const uint8_t* data) {
    uint64_t result = 0;
    for (int i = 0; i < 8; i++) {
        result |= static_cast<uint64_t>(data[i]) << (i * 8);
    }
    return result;
}

} // namespace InternalEngine